pkg/ebpf/dns_stats.go

```diff
@@ -0,0 +1,62 @@
+package ebpf
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"sync"
+)
+
+type dnsStats struct {
+	replies uint32
+	// More stats like latency, error, etc. will be added here later
+}
+
+type dnsData struct {
+	stats             dnsStats
+	lastTransactionID uint16
+}
+
+type dnsKey struct {
+	serverIP   util.Address
+	clientIP   util.Address
+	clientPort uint16
+	protocol   ConnectionType
+}
+
+type dnsStatKeeper struct {
+	mux  sync.Mutex
+	data map[dnsKey]dnsData
+}
+
+func newDNSStatkeeper() *dnsStatKeeper {
+	return &dnsStatKeeper{
+		data: make(map[dnsKey]dnsData),
+	}
+}
+
+func (d *dnsStatKeeper) IncrementReplyCount(key dnsKey, transactionID uint16) {
+	d.mux.Lock()
+	defer d.mux.Unlock()
+	dnsData := d.data[key]
+
+	// For local DNS traffic, sometimes the same reply packet gets processed by the
+	// snooper multiple times. This check avoids double counting in that scenario.
+	if dnsData.lastTransactionID == transactionID {
```

Comment:
If the same packet can be processed more than once, is it possible to for other replies to be interleaved in? For example, `p1, p2, p1, p1` ?

Reply:
One possible reason for duplication of the local DNS traffic is the [following:](https://stackoverflow.com/questions/9196791/duplicate-udp-packets-how-often-it-happens) 

> If the socket in the destination host is bound to INADDR_ANY and if the source broadcasts the UDP packets then there is possibility to receive duplicate packets

I guess in that scenario it is technically possible for packets to be interleaved though I have not been able to reproduce it. To take care of that possibility, we can keep a set of last N transaction IDs and check against that set.
However, this duplication happens only for local DNS traffic which we discard by default. In the rare case that we enable it, checking against the last transaction ID would take care of most if not all of the cases. So the code ignores the highly unlikely case of interleaved duplicate packets in exchange for faster checks in the common case. I will add a comment to that effect. 

If the tradeoff does not make sense or if there is a way to avoid the tradeoff, I would be happy to listen to other ideas.

Reply:
I think thats fine, thanks for explaining.